pkg/encryption/crypter.go

```diff
@@ -0,0 +1,205 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package encryption
+
+import (
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"encoding/binary"
+	"io"
+	"time"
+
+	"github.com/pingcap/kvproto/pkg/encryptionpb"
+	"github.com/tikv/pd/pkg/errs"
+)
+
+const (
+	ivLengthCTR = 16
+	ivLengthGCM = 12
+)
+
+// CheckEncryptionMethodSupported check whether the encryption method is currently supported.
+// This is to handle future extension to encryption methods on kvproto side.
+func CheckEncryptionMethodSupported(method encryptionpb.EncryptionMethod) error {
+	switch method {
+	case encryptionpb.EncryptionMethod_AES128_CTR:
+		return nil
+	case encryptionpb.EncryptionMethod_AES192_CTR:
+		return nil
+	case encryptionpb.EncryptionMethod_AES256_CTR:
+		return nil
+	default:
+		name, ok := encryptionpb.EncryptionMethod_name[int32(method)]
+		if ok {
+			return errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(name)
+		}
+		return errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(int32(method))
+	}
+}
+
+// KeyLength return the encryption key length for supported encryption methods.
+func KeyLength(method encryptionpb.EncryptionMethod) (int, error) {
+	switch method {
+	case encryptionpb.EncryptionMethod_AES128_CTR:
+		return 16, nil
+	case encryptionpb.EncryptionMethod_AES192_CTR:
+		return 24, nil
+	case encryptionpb.EncryptionMethod_AES256_CTR:
+		return 32, nil
+	default:
+		name, ok := encryptionpb.EncryptionMethod_name[int32(method)]
+		if ok {
+			return 0, errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(name)
+		}
+		return 0, errs.ErrEncryptionInvalidMethod.GenWithStackByArgs(int32(method))
+	}
+}
+
+// IvCtr represent IV bytes for CTR mode.
+type IvCtr []byte
+
+// IvGcm represent IV bytes for GCM mode.
+type IvGcm []byte
+
+func newIV(ivLength int) ([]byte, error) {
+	iv := make([]byte, ivLength)
+	n, err := io.ReadFull(rand.Reader, iv)
+	if err != nil {
+		return nil, errs.ErrEncryptionGenerateIV.Wrap(err).GenWithStackByArgs()
+	}
+	if n != ivLength {
+		return nil, errs.ErrEncryptionGenerateIV.GenWithStack(
+			"iv length exepcted %d vs actual %d", ivLength, n)
+	}
+	return iv, nil
+}
+
+// NewIvCtr randomly generate an IV for CTR mode.
+func NewIvCtr() (IvCtr, error) {
+	return newIV(ivLengthCTR)
+}
+
+// NewIvGcm randomly generate an IV for GCM mode.
+func NewIvGcm() (IvGcm, error) {
+	return newIV(ivLengthGCM)
+}
+
+// NewDataKey randomly generate a new data key.
+func NewDataKey(
+	method encryptionpb.EncryptionMethod,
+) (keyID uint64, key *encryptionpb.DataKey, err error) {
+	err = CheckEncryptionMethodSupported(method)
+	if err != nil {
+		return
+	}
+	keyIDBuf := make([]byte, 8)
+	n, err := io.ReadFull(rand.Reader, keyIDBuf)
+	if err != nil {
+		err = errs.ErrEncryptionNewDataKey.Wrap(err).GenWithStack(
+			"fail to generate data key id")
+		return
+	}
+	if n != 8 {
```

Comment:
I think we need `const uint64Size = 8`

Reply:
Is there a Go lang constant that I can use?

Reply:
`unsafe.Sizeof(uint64(0))`

Reply:
addressed in #2976